Make this a pip-installable package

fmauch opened the pull request:
This implements #9 

Please keep discussions about whether this should be implemented in #9 while keeping technical discussions about the actual implementations here.

fmauch commented:
@chriseichmann could you give it a go?

It should be installable using

```
pip install --user -e .
```

Then `catmux_create_session` should be available given that your `PATH` contains `~/.local/bin`.

chriseichmann commented:
I've already started experimenting :-)
Can you give me permission to write to the plain_python branch?

chriseichmann commented:
I can install this branch and run the `create_session` script with the example session yaml.
The installed script is called `create_session`, maybe renaming it would be good.

chriseichmann commented:
nvm.. the script is correctly named

fmauch commented:
One question is what to do with the example files. Especially the tmux config. I would suggest to have it into the installed package folder and use it if no tmux config is present and no tmux config is passed to the instance.

chriseichmann commented:
I think the example config can stay in its current place. It's sufficient for a quick "hello world"-style test from inside the repository.

Regarding the tmux config I agree to make it the default if no other config is found. With catmux being no longer a catkin package, it is no longer locatable via rospack.

fmauch commented on README.md:
```diff
@@ -71,18 +71,39 @@ optional arguments:
+administration or other development tasks.
+
+Catmux no longer has the `package://<your_package_name>` lookup capabilities.
+To archive the same outcome, simply use rospack:
```
```suggestion
To achieve the same outcome, simply use rospack:
```

fmauch commented:
I rebased the branch on the current master as there were license conflicts introduced a while back.

07d5634 should be tested and then I think we are good to finish this.

I'd like to update the README such that it doesn't suggest to use the package root directory but the installed files. I would expect users to not download the source code, but simply install it using pip over pypi once it is released.

fmauch commented:
I tested f8a171b and it works as expected. This way, users without a custom tmux configuration will get this package's default config.

I installed the package using `pip install --user .` and `sudo pip install .`

In the first case the files ended up in `~/.local/catmux`, in the second case in `/var/catmux`. I'm not sure whether that's the best way to do things, though...

chriseichmann commented:
The goal is to have catmux be installed via PyPI, right? In that spirit, `.local` seems to be a good place. Personally, I prefer to let users handle their packages themself using `pip install --user`.

fmauch commented:
This would now be ready from my side. Please review once more @chriseichmann

chriseichmann commented on setup.py:
```diff
@@ -1,12 +1,29 @@
     packages=['catmux'],
-    scripts=['script/create_session'],
-    package_dir={'': 'src'}
+    data_files=[('share/catmux', ['etc/tmux_default.conf',
+                            'etc/example_session.yaml',
+                            'etc/readme_tmux.txt'])],
```
```suggestion
    data_files=[('share/catmux', ['etc/tmux_default.conf',
                                  'etc/example_session.yaml',
                                  'etc/readme_tmux.txt'])],
```

chriseichmann commented on README.md:
```diff
@@ -38,25 +38,29 @@ project and I definitely wanted something that saves the config per project and
 
 ### Running the example the most simple way
-to simply run the example, type
+After installation, you can run a simple example by calling the following command in the
+repository's root directory.
```
```suggestion
After installation, you can run a simple example by calling one of the following commands.
```

chriseichmann commented on README.md:
```diff
@@ -71,18 +75,42 @@ optional arguments:
 
 ### Full blown example
-To make use of all catmux features, run the following example command:
+To make use of all catmux features, run the following example command, again from the root
+directory:
```
```suggestion
To make use of all catmux features, run one of the following example commands:
```

chriseichmann commented on README.md:
````diff
@@ -71,18 +75,42 @@ optional arguments:
+directory:
 ```
-rosrun catmux create_session package://catmux/etc/example_session.yaml --tmux_config package://catmux/etc/tmux_default.conf --session_name example_session --overwrite show_layouts=True,replacement_param="new catmux user"
+catmux_create_session ${HOME}/.local/share/catmux/example_session.yaml \
+  --tmux_config {HOME}/.local/share/catmux/tmux_default.conf \
+  --session_name example_session \
+  --overwrite show_layouts=True,replacement_param="new catmux user"
````
```suggestion
# In case you used the `--user` flag during the install process
catmux_create_session ${HOME}/.local/share/catmux/example_session.yaml \
  --tmux_config {HOME}/.local/share/catmux/tmux_default.conf \
  --session_name example_session \
  --overwrite show_layouts=True,replacement_param="new catmux user"
  
  
# In case you used `sudo pip install ...` for installation
catmux_create_session $(python3 -c "import sys; print(sys.prefix)")/share/catmux/example_session.yaml \
  --tmux_config $(python3 -c "import sys; print(sys.prefix)")/share/catmux/tmux_default.conf \
  --session_name example_session \
  --overwrite show_layouts=True,replacement_param="new catmux user"
```

chriseichmann replied:
Using `--user` the full-blown example from the README works.

chriseichmann reviewed (commented):
Weird, `sudo pip install .` installs to `/usr/local/share/catmux` for me but `import sys; print(sys.prefix)`returns `/usr`

Therefore, the examples in the readme do not work for me

chriseichmann commented on script/catmux_create_session:
```diff
@@ -103,8 +88,19 @@ def main():
+        tmux_config = os.path.expanduser("~/.tmux.conf")
+    elif os.path.exists("/etc/tmux.conf"):
+        tmux_config = "/etc/tmux.conf"
+    elif os.path.exists(os.path.join(site.USER_BASE, 'catmux', 'tmux_default.conf')):
+        tmux_config = os.path.join(site.USER_BASE, 'catmux', 'tmux_default.conf')
```
```suggestion
    elif os.path.exists(os.path.join(site.USER_BASE, 'share', 'catmux', 'tmux_default.conf')):
        tmux_config = os.path.join(site.USER_BASE, 'share', 'catmux', 'tmux_default.conf')
```

This should fix the defaulting to the default tmux config if no config is provided or present for the `--user` case

chriseichmann reviewed (commented):
A possible fix for the tmux config lookup using `pip install --user .`
Currently, the default config is expected at `~/.local/catmux/` but installed at `~/.local/share/catmux/`

chriseichmann commented:
I tested both installation "types" without a tmux config present and suggested a fix for the `--user` installation.
The installation path with `sudo pip` seems to vary, therefore, your proposed example from the readme does not work for me.

fmauch commented:
> Weird, `sudo pip install .` installs to `/usr/local/share/catmux` for me but `import sys; print(sys.prefix)`returns `/usr`
> 
> Therefore, the examples in the readme do not work for me

I was feeling concerned about this...

chriseichmann commented:
`import catmux; catmux.__file__.split("lib")[0]` does the trick, but isn't pretty

chriseichmann commented:
Or piggyback of on `pip show catmux` somehow. This would get rid of the distinction between `sudo` and `--user` installation at least.

fmauch commented:
Yes, Something similar could be packed into a submodule. I'll do it later today.

On November 25, 2020 8:06:43 AM GMT+01:00, chriseichmann <notifications@github.com> wrote:
>`import catmux; catmux.__file__.split("lib")[0]` does the trick, but
>isn't pretty
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/fmauch/catmux/pull/10#issuecomment-733511830

fmauch commented:
> `import catmux; catmux.__file__.split("lib")[0]` does the trick, but isn't pretty

Implemented in dc58b7f

Could you test it once more @chriseichmann?

chriseichmann commented:
Funny thing: if you are inside the repo root directory after installation, `python3 -m catmux.prefix` will return a different path as anywhere else:
```bash
~$ cd ~/git/catmux/
~/git/catmux$ python3 -m catmux.prefix
/home/eichmann/git/catmux/catmux/prefix.py
~/git/catmux$ cd
~$ python3 -m catmux.prefix
/home/eichmann/.local/
```
I will change the wording of the readme slightly.

fmauch commented:
Yes, same happens if you install it using `pip install -e`

chriseichmann commented:
`pip install -e` is a special case and I think we don't need to handle this one explicitly in the README.
We should add the dependencies to the setup.py, like `yaml` and `future` and than we are good to go

fmauch commented:
I'd actually like to get rid of future and fully migrate towards python3 as a requirement. As the `setup.py` already requires `python>=3.6` I would pack this into this PR.

fmauch commented:
@chriseichmann I've cleaned up the branch and I think it is ready to merge once you approve this.

chriseichmann commented:
Okay, I think the PR is ready to be merged.
